packages/sitecore-jss-react/src/components/ErrorBoundary.tsx

```diff
@@ -0,0 +1,74 @@
+import React, { ReactNode, Suspense } from 'react';
+import { ComponentRendering } from '@sitecore-jss/sitecore-jss/layout';
+import { withSitecoreContext } from '../enhancers/withSitecoreContext';
+import { SitecoreContextValue } from './SitecoreContext';
+
+type CustomErrorComponentProps = {
+  [prop: string]: unknown;
+};
+
+export type ErrorBoundaryProps = {
+  children: ReactNode;
+  customErrorComponent?:
+    | React.ComponentClass<CustomErrorComponentProps>
+    | React.FC<CustomErrorComponentProps>;
+  rendering?: ComponentRendering;
+  sitecoreContext: SitecoreContextValue;
+  componentLoadingMessage?: string;
+  type: string;
+};
+
+class ErrorBoundary extends React.Component<ErrorBoundaryProps> {
+  state: { error: Error };
+
+  constructor(props: any) {
+    super(props);
+    this.state = { error: null };
+  }
+
+  static getDerivedStateFromError(error: Error) {
+    return { error: error };
+  }
+
+  componentDidCatch(error: Error, errorInfo: any) {
+    console.error({ error, errorInfo });
+  }
+
+  render() {
+    if (this.state.error) {
+      if (this.props.sitecoreContext?.pageEditing) {
+        return (
+          <div>
+            <div className="sc-jss-placeholder-error">
+              A rendering error occurred in component <em>{this.props.rendering?.componentName}</em>
+              <br />
+              Error: <em>{this.state.error.message}</em>
+            </div>
+          </div>
+        );
+      } else {
+        if (this.props.customErrorComponent) {
+          return <this.props.customErrorComponent error={this.state.error} />;
+        } else {
+          return (
+            <div>
+              <div className="sc-jss-placeholder-error">
+                A rendering error occurred in component <em>{this.state.error.message}</em>
+              </div>
+            </div>
+          );
+        }
+      }
+    }
```

Comment:
Should probably use the `customErrorComponent` in edit mode as well ... but do we need a difference between edit/normal mode? We don't make that distinction in our error handling at the placeholder level (code [here](https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-react/src/components/Placeholder.tsx#L74-L84), docs [here](https://doc.sitecore.com/xmc/en/developers/jss/217/jss-xmc/working-with-placeholders-in-a-jss-next-js-app.html#error-components)).

Also, is there any reason to keep the error handling at the placeholder level? Seems like this effectively supersedes that.

Reply:
Hey @ambrauer,
we had a discussion regarding this on Friday.
I have reworked it with the following logic:
- if custom error component has been provided - show it.
- if we are in development mode or in editing mode - show a message with additional information - error's message and the name of the faulty component (could be useful for troubleshooting).
- if error is in production and not editing mode - show a generic message for the final user, without details.
What's your opinion regarding the above?

And yes, I guess we could remove the error handling on placeholder level.

Reply:
@yavorsk Makes sense, just one last thought about that last one...
 
 > if error is in production and not editing mode - show a generic message for the final user, without details.
 
 I know this is how Placeholder-level error handling is working, but I'm wondering if we should render anything at all in this case? If an error occurs on your production website, wouldn't it be preferrable to hide the erroneous component and limit the error details to the console/logging?

Reply:
That also makes sense. I'll ask for Kusuma's opinion regarding this.

Reply:
@aweber1, @yavorsk 
If the error affects the user experience, we should show error on whats happening otherwise they won't know right?
For example, when I click on something it meant to take me to another component but due to error if we don't show anything user won't know what happened.
Am I missing something?

Reply:
@rangkusu True as well. I think it really depends on the component and how it's used on a page.
@yavorsk I say let's leave it as-is (consistent with Placeholder-level error handling). If a customer _did_ want to hide it when in an error state on production, they could easily accomplish this with a custom `errorComponent`.